content/documentation/langium/content.md

```diff
@@ -1,13 +1,23 @@
 +++
 fragment = "content"
 title = "Langium Integration"
-weight = 150
+weight = 160
 [sidebar]
   sticky = true
 +++
 
-TODO say what langium is and link to it
+To extend the Model Hub with your specific language, you need to provide a `ModelServiceContribution`.
+A model service contribution can provide a dedicated Model Service API and extend the persistence and validation capabilities of the Model Hub.
 
-TODO describe briefly how a language is defined in langium
+One of the core principles in our Model Hub architecture is re-use and we therefore aim to re-use as much of the language infrastructure and support that Langium generates for us.
+This is reflected in several design decisions:
+- Since the language server that contains the modules with all the languages already starts in a dedicated process, we will start our own Model Hub server in the same process to ease access.
+- We are re-using the dependency injection framework from Langium to bind our own Model Hub-specific services, such as the core model hub implementation, the model manager that holds the model storage, the overall command stack and the model subscriptions or the validation service that ensures that all custom validations are run on each model.
 
-TODO describe in more detail how to register the respective Langium services for a language to the model hub to fulfill the model hub API for a language contribution and mention which of the Model Hub capabilities are then covered generically
+The bridge that connects the Model Hub world with the Langium world is our generic _EMF.cloud Model Hub Langium integration library_.
+That library has two main components:
+1. The Abstract Syntax Tree (AST) server that serves as a facade to access and update semantic models from the Langium language server as a non-LSP client. It provides a simple open-request-update-save/close lifecycle for documents and their semantic model.
+2. A converter between the Langium-based AST model and the client language model. The biggest difference between those two models is that the client language model needs to be serializable as we intend to send it to the model hub client which might run in a differenct process. This core work of this transformation is the resolution of cycles and the proper representation of cross references so that when we get a language model back from the client we can restore a full Langium-based AST model again, i.e., to have a full bi-directional transformation when it comes to the semantic model.
```

Comment:
> needs to be serializable

I would clarify this as: `needs to be serializable to JSON` (any Langium document is serializable, since they are text documents in the first place). The ModelHub explicitly works with JSON Documents (because we use Json Patches to describe changes).

Reply:
The problem is that Langium documents are actually not serializable because they do contain cycles and structures that may not be serializable such as URIs. However, if you do not have any cycles, any JavaScript object is serializable to JSON and that is actually what we apply our JSON patches on.

I just want to make sure that we have a very clear distinction between the JSON-based grammar, the abstract syntax tree (which are just JavaScript objects with cycles) and the client language model (which are also just any JavaScript objects without cycles). I can add the `to JSON` part, I just want to avoid any confusion about this.

Reply:
> if you do not have any cycles, any JavaScript object is serializable to JSON and that is actually what we apply our JSON patches on.

Not quite, actually. As you mentioned, cycles is the most obvious issue, but even regular references are poorly handled by JSON while being valid in JS. There are also other differences ("undefined" does not exist in JSON for example).

> I just want to make sure that we have a very clear distinction between the JSON-based grammar

That's specific to the Coffee Editor, right? Or do we expect all "EMF Cloud" grammars to be JSON-like?

In a more general case, what should it look like if a language is absolutely not JSON-like, but still wants to interact with the ModelManager? I think it should have a model-to-JSON and JSON-to-model conversion, taking care of serializing and de-serializing cross-references? (similar to what we had for EMF-to-JSON?). So in that regard, what we really want in the general Model Hub use case is a "pure JSON" document, where x-refs are serialized as simple strings (or simple objects with type/uri properties)?

I didn't really look at how cross references are handled in practice in the Coffee Editor, so I'm not sure exactly what we expect and how this should be documented.

Reply:
Right, so what we have is:
- The serialization format defined through the grammar (e.g., JSON-based, YAML-based or completely custom)
- The abstract syntax tree model that Langium parses out of documents that conform to the serialization format. This will always be a set of JavaScript objects that cannot be serialized automatically.
- The JSON model that we use in Model Hub based on the AST from Langium but without cycles and with pure JSON as we need to support JSON Patches and serialization.

The serialization format can almost be neglected as we almost exclusively deal with the AST and the JSON model. I just want to avoid the confusion between the JSON model we use in Model Hub and the JSON-based grammar model that defines the serialization model.

I added another sentence to clarify this.